chasm/tree.go

```diff
@@ -1262,30 +1307,91 @@ func (n *Node) closeTransactionHandleRootLifecycleChange() (bool, error) {
 }
 
 func (n *Node) closeTransactionForceUpdateVisibility(
-	nextVersionedTransition *persistencespb.VersionedTransition,
+	rootLifecycleChanged bool,
 ) error {
+	var visibilityNode *Node
 	for _, child := range n.children {
 		componentAttr := child.serializedNode.Metadata.GetComponentAttributes()
-		if componentAttr == nil ||
-			componentAttr.Type != visibilityComponentFqType ||
-			transitionhistory.Compare(child.serializedNode.Metadata.LastUpdateVersionedTransition, nextVersionedTransition) == 0 {
-			// Skip force update is the node is not a visibility component or already updated in this transition.
+		if componentAttr == nil {
 			continue
 		}
 
-		mutableContext := NewMutableContext(context.Background(), n)
-		visComponent, err := child.Component(mutableContext, ComponentRef{})
-		if err != nil {
-			return err
+		if child.valueState == valueStateNeedSerialize {
+			if rc, ok := n.registry.componentFor(child.value); ok && rc.fqType() == visibilityComponentFqType {
+				visibilityNode = child
+				break
+			}
+		} else if componentAttr.Type == visibilityComponentFqType {
+			visibilityNode = child
+			break
 		}
 
-		visibility, ok := visComponent.(*Visibility)
-		if !ok {
-			return serviceerror.NewInternalf("expected visibility component for component with type %s, but got %T", visibilityComponentFqType, visComponent)
+	}
+	if visibilityNode == nil {
+		return nil
+	}
+
+	needUpdate := rootLifecycleChanged
+
+	mutableContext := NewMutableContext(context.Background(), n)
+	immutableContext := mutableContext.ToImmutable()
+
+	visComponent, err := visibilityNode.Component(immutableContext, ComponentRef{})
+	if err != nil {
+		return err
+	}
+
+	visibility, ok := visComponent.(*Visibility)
+	if !ok {
+		return serviceerror.NewInternalf("expected visibility component for component with type %s, but got %T", visibilityComponentFqType, visComponent)
+	}
+
+	rootComponent, err := n.Component(immutableContext, ComponentRef{})
+	if err != nil {
+		return err
+	}
+
+	saProvider, ok := rootComponent.(VisibilitySearchAttributesProvider)
+	if ok {
+		newSearchAttributes := saProvider.SearchAttributes(immutableContext)
+		if !maps.Equal(n.currentSearchAttributes, newSearchAttributes) {
```

Comment:
This doesn't work with keyword lists:

```
func TestFoo(t *testing.T) {
	m1 := map[string]any{"a": []string{"a", "b", "c"}}
	m2 := map[string]any{"a": []string{"a", "b", "c"}}
	require.True(t, maps.Equal(m1, m2))
}


// panic: runtime error: comparing uncomparable type []string [recovered, repanicked
```

Reply:
ah thanks for catching! Missed keyword list case, will find a way to fix.

Reply:
Resolving the comment as it no longer applies with the new approach.